File: backend/services/game_service.py
from sqlalchemy.orm import Session
from sqlalchemy.orm.attributes import flag_modified
from typing import List, Optional, Dict
from models.game_state import UserGameState, GameHistory, UserStatistics
from models.user import User
from schemas.game_schema import Card
from services.card_service import generate_random_cards


def get_or_create_game_state(db: Session, user: User, hand_size: int = 5) -> UserGameState:
    """
    Get or create game state for a user.

    Note: hand_size represents game difficulty level (3=Small, 4=Medium, 5=Large),
    NOT the number of hand positions. Hand is always 5 positions.
    """
    game_state = db.query(UserGameState).filter(UserGameState.user_id == user.id).first()

    if not game_state:
        # Validate game_size (difficulty level)
        if hand_size < 3 or hand_size > 5:
            hand_size = 5  # Default to 5 (Large) if invalid

        # Create new game state with empty hand (always 5 slots)
        game_state = UserGameState(
            user_id=user.id,
            hand=[None] * 5,  # Always 5 positions
            hand_size=hand_size,  # This is difficulty level (3-5)
            deck=[],
            discard_pile=[]
        )
        db.add(game_state)

        # Also create statistics
        stats = UserStatistics(user_id=user.id)
        db.add(stats)

        db.commit()
        db.refresh(game_state)

    return game_state


def get_deck_composition(game_state: UserGameState) -> Dict[str, int]:
    """Calculate deck composition (for display purposes)"""
    composition = {}
    for card in game_state.deck:
        if card:
            card_type = f"{card.get('Type', 'Unknown')} ({card.get('color', 'Unknown')})"
            composition[card_type] = composition.get(card_type, 0) + 1
    return composition


def update_user_hand(db: Session, user: User, new_hand: List[Optional[Dict]]) -> UserGameState:
    """Update user's hand with new cards"""
    game_state = get_or_create_game_state(db, user)

    # Validate hand is always 5 positions
    if len(new_hand) != 5:
        raise ValueError("Hand must have exactly 5 positions")

    # Update hand
    game_state.hand = new_hand
    flag_modified(game_state, "hand")
    db.commit()
    db.refresh(game_state)

    # Log action
    log_game_action(db, user.id, "update_hand", {"hand": new_hand})

    return game_state


def play_card_from_hand(db: Session, user: User, hand_position: int, discard_positions: Optional[List[int]] = None) -> tuple[UserGameState, Optional[List[Card]], bool, List[int], Optional[Dict]]:
    """
    Play a card from hand.
    Returns: (updated_game_state, drawn_cards, auto_placed, placed_positions, curse_data)
    - drawn_cards: Cards that were drawn (None if no cards drawn)
    - auto_placed: True if cards were placed, False if user needs to choose placement
    - placed_positions: List of hand positions where cards were placed
    - curse_data: Curse card data if a curse was played (None otherwise)
    """
    game_state = get_or_create_game_state(db, user)

    if hand_position < 0 or hand_position >= game_state.hand_size:
        raise ValueError("Invalid hand position")

    card = game_state.hand[hand_position]
    if not card:
        raise ValueError("No card at this position")

    # Check if it's a special card
    drawn_cards = None
    auto_placed = True  # Default to True for regular cards
    placed_positions = []
    curse_data = None

    if card.get("Type") == "Curse":
        # Handle curse card
        casting_cost = card.get("casting_cost", {})
        discard_count = casting_cost.get("discard", 0)

        if discard_count > 0:
            # Validate discard positions
            if not discard_positions or len(discard_positions) != discard_count:
                raise ValueError(f"Must select {discard_count} card(s) to discard")

            # Check for duplicate positions
            if len(discard_positions) != len(set(discard_positions)):
                raise ValueError("Cannot select the same card position multiple times")

            # Validate discard positions
            for pos in discard_positions:
                if pos < 0 or pos >= game_state.hand_size or pos == hand_position:
                    raise ValueError(f"Invalid discard position: {pos}")
                if not game_state.hand[pos]:
                    raise ValueError(f"No card at position {pos}")

            # Discard selected cards
            for pos in discard_positions:
                game_state.discard_pile.append(game_state.hand[pos])
                game_state.hand[pos] = None

        # Move curse card to discard pile
        game_state.discard_pile.append(card)
        game_state.hand[hand_position] = None
        flag_modified(game_state, "hand")
        flag_modified(game_state, "discard_pile")

        # Store curse data to return to frontend
        curse_data = card

    elif card.get("name") == "Discard 1 Draw 2":
        if not discard_positions or len(discard_positions) != 1:
            raise ValueError("Must select 1 card to discard")
        drawn_cards, auto_placed, placed_positions = handle_discard_draw(db, game_state, user, hand_position, discard_positions, 1, 2)

    elif card.get("name") == "Discard 2 Draw 3":
        if not discard_positions or len(discard_positions) != 2:
            raise ValueError("Must select 2 cards to discard")
        drawn_cards, auto_placed, placed_positions = handle_discard_draw(db, game_state, user, hand_position, discard_positions, 2, 3)

    else:
        # Regular card - just remove from hand
        game_state.discard_pile.append(card)
        game_state.hand[hand_position] = None
        flag_modified(game_state, "hand")
        flag_modified(game_state, "discard_pile")

    db.commit()
    db.refresh(game_state)

    # Update statistics
    update_statistics(db, user.id, cards_played=1)

    # Log action
    log_game_action(db, user.id, "play_card", {
        "card": card,
        "position": hand_position,
        "discard_positions": discard_positions
    })

    return game_state, drawn_cards, auto_placed, placed_positions, curse_data


def handle_discard_draw(db: Session, game_state: UserGameState, user: User,
                       played_position: int, discard_positions: List[int],
                       discard_count: int, draw_count: int) -> tuple[List[Card], bool, List[int]]:
    """
    Handle special 'Discard X Draw Y' cards.
    Returns: (drawn_cards, auto_placed, placed_positions)
    - drawn_cards: The newly drawn cards
    - auto_placed: True if cards were automatically placed, False if user needs to choose placement
    - placed_positions: List of hand positions where cards were placed (empty if not auto-placed)
    """

    # Check for duplicate positions
    if len(discard_positions) != len(set(discard_positions)):
        raise ValueError("Cannot select the same card position multiple times")

    # Validate discard positions
    for pos in discard_positions:
        if pos < 0 or pos >= game_state.hand_size or pos == played_position:
            raise ValueError(f"Invalid discard position: {pos}")
        if not game_state.hand[pos]:
            raise ValueError(f"No card at position {pos}")

    # Discard cards
    for pos in discard_positions:
        game_state.discard_pile.append(game_state.hand[pos])
        game_state.hand[pos] = None

    # Remove the played card itself
    game_state.discard_pile.append(game_state.hand[played_position])
    game_state.hand[played_position] = None

    # Mark as modified so SQLAlchemy tracks the changes
    flag_modified(game_state, "hand")
    flag_modified(game_state, "discard_pile")

    # Draw new cards filtered by game difficulty
    new_cards = generate_random_cards(draw_count, game_state.hand_size)

    # Update statistics
    update_statistics(db, user.id, cards_drawn=len(new_cards))

    # Log action
    log_game_action(db, user.id, "draw_cards", {"count": len(new_cards)})

    # Check if there's enough empty space in hand for the drawn cards
    empty_slots = [i for i, card in enumerate(game_state.hand) if card is None]

    if len(empty_slots) >= len(new_cards):
        # Automatically place cards in empty slots
        placed_positions = []
        for i, card in enumerate(new_cards):
            position = empty_slots[i]
            game_state.hand[position] = card.dict()
            placed_positions.append(position)
        flag_modified(game_state, "hand")
        return new_cards, True, placed_positions
    else:
        # Not enough space - user must choose what to discard
        return new_cards, False, []


def log_game_action(db: Session, user_id: int, action_type: str, action_data: Dict):
    """Log a game action to history"""
    history_entry = GameHistory(
        user_id=user_id,
        action_type=action_type,
        action_data=action_data
    )
    db.add(history_entry)
    db.commit()


def place_pending_cards(db: Session, user: User, cards_to_place: List[Dict], discard_positions: List[int]) -> UserGameState:
    """
    Place pending cards into the hand, discarding cards as necessary.
    This is called when drawn cards couldn't be auto-placed due to lack of space.

    Args:
        cards_to_place: The cards that need to be placed into the hand
        discard_positions: Positions of cards to discard to make room
    """
    game_state = get_or_create_game_state(db, user)

    # Validate that we're discarding the right number of cards
    empty_slots = sum(1 for card in game_state.hand if card is None)
    cards_needed = len(cards_to_place)
    additional_discards_needed = cards_needed - empty_slots

    if additional_discards_needed < 0:
        raise ValueError("There's already enough space in the hand")

    if len(discard_positions) != additional_discards_needed:
        raise ValueError(f"Must discard exactly {additional_discards_needed} card(s) to make room")

    # Validate discard positions
    if len(discard_positions) != len(set(discard_positions)):
        raise ValueError("Cannot select the same card position multiple times")

    for pos in discard_positions:
        if pos < 0 or pos >= game_state.hand_size:
            raise ValueError(f"Invalid discard position: {pos}")
        if not game_state.hand[pos]:
            raise ValueError(f"No card at position {pos}")

    # Discard the selected cards
    for pos in discard_positions:
        game_state.discard_pile.append(game_state.hand[pos])
        game_state.hand[pos] = None

    # Now place the new cards in empty slots
    empty_slots_indices = [i for i, card in enumerate(game_state.hand) if card is None]

    if len(empty_slots_indices) < len(cards_to_place):
        raise ValueError("Not enough space even after discarding")

    for i, card in enumerate(cards_to_place):
        game_state.hand[empty_slots_indices[i]] = card

    # Mark as modified
    flag_modified(game_state, "hand")
    flag_modified(game_state, "discard_pile")

    db.commit()
    db.refresh(game_state)

    # Log action
    log_game_action(db, user.id, "place_pending_cards", {
        "cards_placed": len(cards_to_place),
        "discarded_positions": discard_positions
    })

    return game_state


def update_statistics(db: Session, user_id: int, cards_drawn: int = 0, cards_played: int = 0, games_completed: int = 0):
    """Update user statistics"""
    stats = db.query(UserStatistics).filter(UserStatistics.user_id == user_id).first()

    if stats:
        stats.total_cards_drawn += cards_drawn
        stats.total_cards_played += cards_played
        stats.games_completed += games_completed
        db.commit()


def update_hand_size(db: Session, user: User, new_game_size: int) -> UserGameState:
    """
    Update the game difficulty level (game_size) for a user's game state.

    Note: This does NOT resize the hand array - hand always stays at 5 positions.
    The hand_size field stores the difficulty level (3=Small, 4=Medium, 5=Large).
    """
    # Validate game_size (difficulty level)
    if new_game_size < 3 or new_game_size > 5:
        raise ValueError("Game size must be between 3 and 5")

    game_state = get_or_create_game_state(db, user)
    old_game_size = game_state.hand_size

    if old_game_size == new_game_size:
        return game_state

    # Update game_size field (hand stays at 5 positions)
    game_state.hand_size = new_game_size
    db.commit()
    db.refresh(game_state)

    # Log action
    log_game_action(db, user.id, "update_game_size", {
        "old_size": old_game_size,
        "new_size": new_game_size
    })

    return game_state

File: api/services/gameService.js
import { PrismaClient } from '@prisma/client';
import { buildDeck, drawFromDeck, consumeFromDeck } from './cardService.js';

const prisma = new PrismaClient();
const HAND_SIZE = parseInt(process.env.HAND_SIZE) || 5;

export async function getOrCreateGameState(userId, gameSize = 5) {
  let gameState = await prisma.userGameState.findUnique({
    where: { userId },
  });

  if (!gameState) {
    // Initialize empty hand with null positions
    gameState = await prisma.userGameState.create({
      data: {
        userId,
        hand: Array(HAND_SIZE).fill(null),
        gameSize,
        deck: [],
        discardPile: [],
      },
    });
  } else {
    // Check if hand size needs to be adjusted to match HAND_SIZE environment variable
    const currentHandSize = gameState.hand.length;
    if (currentHandSize !== HAND_SIZE) {
      const newHand = [...gameState.hand];

      if (currentHandSize < HAND_SIZE) {
        // Add empty slots to match new size
        while (newHand.length < HAND_SIZE) {
          newHand.push(null);
        }
      } else {
        // Remove slots from the end (keep first HAND_SIZE cards)
        newHand.splice(HAND_SIZE);
      }

      // Update the game state with the resized hand
      gameState = await prisma.userGameState.update({
        where: { userId },
        data: { hand: newHand },
      });
    }
  }

  return gameState;
}

export async function updateUserHand(userId, newHand) {
  if (!Array.isArray(newHand) || newHand.length !== HAND_SIZE) {
    throw new Error(`Hand must be an array of exactly ${HAND_SIZE} positions`);
  }

  const gameState = await prisma.userGameState.findUnique({ where: { userId } });
  if (!gameState) throw new Error('Game state not found');

  let deck = Array.isArray(gameState.deck) ? gameState.deck : [];
  const hasPending = deck.some(c => c._pending);

  if (hasPending) {
    const oldHand = gameState.hand;

    // Count card IDs in old hand to detect genuinely new cards (not rearranged)
    const oldIdCounts = {};
    for (const card of oldHand) {
      if (card) oldIdCounts[card.id] = (oldIdCounts[card.id] || 0) + 1;
    }
    const newIdCounts = {};
    for (const card of newHand) {
      if (card) newIdCounts[card.id] = (newIdCounts[card.id] || 0) + 1;
    }

    // For each card id appearing more in new hand than old, those extras are genuinely new
    let updatedDeck = [...deck];
    for (const [idStr, newCount] of Object.entries(newIdCounts)) {
      const id = parseInt(idStr);
      const oldCount = oldIdCounts[id] || 0;
      const genuinelyNew = newCount - oldCount;
      for (let k = 0; k < genuinelyNew; k++) {
        // Remove one pending card with this id from the deck
        const pendingIdx = updatedDeck.findIndex(c => c._pending && c.id === id);
        if (pendingIdx !== -1) {
          updatedDeck.splice(pendingIdx, 1);
        }
      }
    }

    // Clear remaining pending markers (unselected drawn cards return to pool)
    updatedDeck = updatedDeck.map(c => {
      if (c._pending) {
        const { _pending, ...rest } = c;
        return rest;
      }
      return c;
    });

    deck = updatedDeck;
  }

  return await prisma.userGameState.update({
    where: { userId },
    data: { hand: newHand, deck },
  });
}

export async function updateGameSize(userId, newGameSize) {
  if (newGameSize < 3 || newGameSize > 5) {
    throw new Error('Game size must be between 3 and 5');
  }

  // Rebuild deck for the new game size
  const newDeck = buildDeck(newGameSize);

  return await prisma.userGameState.update({
    where: { userId },
    data: { gameSize: newGameSize, deck: newDeck },
  });
}

export async function drawCardsForUser(userId, count) {
  let gameState = await prisma.userGameState.findUnique({ where: { userId } });
  if (!gameState) {
    gameState = await getOrCreateGameState(userId);
  }

  let deck = Array.isArray(gameState.deck) ? gameState.deck : [];

  // Initialize deck on first draw
  if (deck.length === 0) {
    deck = buildDeck(gameState.gameSize);
  }

  // Draw from deck (clears old pending, marks new ones)
  const { drawn, updatedDeck } = drawFromDeck(deck, count);

  await prisma.userGameState.update({
    where: { userId },
    data: { deck: updatedDeck },
  });

  return drawn;
}

export async function playCardFromHand(userId, handPosition, discardPositions = null) {
  const gameState = await prisma.userGameState.findUnique({
    where: { userId },
  });

  if (!gameState) {
    throw new Error('Game state not found');
  }

  const hand = gameState.hand;
  if (handPosition < 0 || handPosition >= HAND_SIZE) {
    throw new Error('Invalid hand position');
  }

  const card = hand[handPosition];
  if (!card) {
    throw new Error('No card at this position');
  }

  let drawnCards = [];
  let autoPlaced = false;
  let placedPositions = [];
  let curseData = null;

  // Handle Curse cards
  if (card.Type === 'Curse') {
    const castingCost = card.casting_cost || {};
    const discardCost = castingCost.discard || 0;

    if (discardCost > 0) {
      if (!discardPositions || discardPositions.length !== discardCost) {
        throw new Error(`Must discard exactly ${discardCost} card(s) for this curse`);
      }

      // Validate discard positions
      for (const pos of discardPositions) {
        if (pos < 0 || pos >= HAND_SIZE || pos === handPosition) {
          throw new Error('Invalid discard position');
        }
        if (!hand[pos]) {
          throw new Error(`No card at position ${pos} to discard`);
        }
      }

      // Move discarded cards to discard pile
      const discardPile = gameState.discardPile;
      for (const pos of discardPositions) {
        discardPile.push(hand[pos]);
        hand[pos] = null;
      }

      await prisma.userGameState.update({
        where: { userId },
        data: { discardPile },
      });
    }

    // Move curse to discard pile
    const discardPile = gameState.discardPile;
    discardPile.push(card);
    hand[handPosition] = null;

    curseData = {
      name: card.name,
      curse_text: card.curse_text,
      casting_cost: card.casting_cost,
    };

    await prisma.userGameState.update({
      where: { userId },
      data: { hand, discardPile },
    });
  }
  // Handle Discard X Draw Y cards
  else if (card.name === 'Discard 1 Draw 2') {
    const result = await handleDiscardDraw(
      userId,
      gameState,
      handPosition,
      discardPositions,
      1,
      2
    );
    drawnCards = result.drawnCards;
    autoPlaced = result.autoPlaced;
    placedPositions = result.placedPositions;
  } else if (card.name === 'Discard 2 Draw 3') {
    const result = await handleDiscardDraw(
      userId,
      gameState,
      handPosition,
      discardPositions,
      2,
      3
    );
    drawnCards = result.drawnCards;
    autoPlaced = result.autoPlaced;
    placedPositions = result.placedPositions;
  }
  // Regular cards (Time Bonus, Action)
  else {
    const discardPile = gameState.discardPile;
    discardPile.push(card);
    hand[handPosition] = null;

    await prisma.userGameState.update({
      where: { userId },
      data: { hand, discardPile },
    });
  }

  // Update statistics
  await updateStatistics(userId, { cardsPlayed: 1 });

  const updatedState = await prisma.userGameState.findUnique({
    where: { userId },
  });

  return {
    gameState: updatedState,
    drawnCards,
    autoPlaced,
    placedPositions,
    curseData,
  };
}

async function handleDiscardDraw(userId, gameState, playedPosition, discardPositions, discardCount, drawCount) {
  const hand = gameState.hand;

  // Validate discard count
  if (!discardPositions || discardPositions.length !== discardCount) {
    throw new Error(`Must discard exactly ${discardCount} card(s)`);
  }

  // Validate discard positions
  for (const pos of discardPositions) {
    if (pos < 0 || pos >= HAND_SIZE || pos === playedPosition) {
      throw new Error('Invalid discard position');
    }
    if (!hand[pos]) {
      throw new Error(`No card at position ${pos} to discard`);
    }
  }

  // Move discarded cards to discard pile
  const discardPile = gameState.discardPile;

  // Add played card to discard
  discardPile.push(hand[playedPosition]);
  hand[playedPosition] = null;

  // Add discarded cards
  for (const pos of discardPositions) {
    discardPile.push(hand[pos]);
    hand[pos] = null;
  }

  // Draw new cards from deck (immediately consumed — no pending step)
  let deck = Array.isArray(gameState.deck) ? gameState.deck : [];
  if (deck.length === 0) {
    deck = buildDeck(gameState.gameSize);
  }
  const { drawn: drawnCards, updatedDeck } = consumeFromDeck(deck, drawCount);

  // Try to auto-place cards
  const emptySlots = hand
    .map((card, idx) => (card === null ? idx : null))
    .filter(idx => idx !== null);

  let autoPlaced = false;
  let placedPositions = [];

  if (emptySlots.length >= drawnCards.length) {
    // Auto-place all drawn cards
    for (let i = 0; i < drawnCards.length; i++) {
      hand[emptySlots[i]] = drawnCards[i];
      placedPositions.push(emptySlots[i]);
    }
    autoPlaced = true;
  }

  await prisma.userGameState.update({
    where: { userId },
    data: { hand, discardPile, deck: updatedDeck },
  });

  return { drawnCards, autoPlaced, placedPositions };
}

export async function placePendingCards(userId, cardsToPlace, discardPositions) {
  const gameState = await prisma.userGameState.findUnique({
    where: { userId },
  });

  if (!gameState) {
    throw new Error('Game state not found');
  }

  const hand = gameState.hand;

  // Validate discard positions
  if (discardPositions.length !== cardsToPlace.length) {
    throw new Error('Must discard same number of cards as placing');
  }

  for (const pos of discardPositions) {
    if (pos < 0 || pos >= HAND_SIZE) {
      throw new Error('Invalid discard position');
    }
    if (!hand[pos]) {
      throw new Error(`No card at position ${pos} to discard`);
    }
  }

  // Discard cards
  const discardPile = gameState.discardPile;
  for (const pos of discardPositions) {
    discardPile.push(hand[pos]);
    hand[pos] = null;
  }

  // Place new cards
  const emptySlots = hand
    .map((card, idx) => (card === null ? idx : null))
    .filter(idx => idx !== null);

  for (let i = 0; i < cardsToPlace.length; i++) {
    if (emptySlots[i] !== undefined) {
      hand[emptySlots[i]] = cardsToPlace[i];
    }
  }

  return await prisma.userGameState.update({
    where: { userId },
    data: { hand, discardPile },
  });
}

export async function logGameAction(userId, actionType, actionData) {
  return await prisma.gameHistory.create({
    data: {
      userId,
      actionType,
      actionData,
    },
  });
}

export async function updateStatistics(userId, updates) {
  const stats = await prisma.userStatistics.findUnique({
    where: { userId },
  });

  if (!stats) {
    return await prisma.userStatistics.create({
      data: {
        userId,
        totalCardsDrawn: updates.cardsDrawn || 0,
        totalCardsPlayed: updates.cardsPlayed || 0,
        gamesCompleted: updates.gamesCompleted || 0,
      },
    });
  }

  return await prisma.userStatistics.update({
    where: { userId },
    data: {
      totalCardsDrawn: { increment: updates.cardsDrawn || 0 },
      totalCardsPlayed: { increment: updates.cardsPlayed || 0 },
      gamesCompleted: { increment: updates.gamesCompleted || 0 },
    },
  });
}

export function getDeckComposition(gameState) {
  const deck = gameState.deck || [];
  const composition = {};

  for (const card of deck) {
    const key = card.color ? `${card.Type} (${card.color})` : card.Type;
    composition[key] = (composition[key] || 0) + 1;
  }

  return composition;
}
